org.lflang/src/org/lflang/graph/InstantiationGraph.java

```diff
@@ -0,0 +1,178 @@
+ *
+ * In addition to the graph, this class keeps track of the instantiations
+ * that induce the dependencies. These can be retrieved using the method
+ * `getInstantiations(Reactor)`.
+ *
+ * @author{Marten Lohstroh <marten@berkeley.edu>}
+ */
+public class InstantiationGraph extends PrecedenceGraph<Reactor> {
+    /**
+     * A mapping from reactors to the sites of their instantiation.
+     */
+    protected final HashMultimap<Reactor, Instantiation> reactorToInstantiation =
+        HashMultimap.<Reactor, Instantiation>create();
+
+    /**
+     * A mapping from reactor classes to their declarations.
+     */
+    protected final HashMultimap<Reactor, ReactorDecl> reactorToDecl =
+        HashMultimap.<Reactor, ReactorDecl>create();
+
+    protected final HashSet<Object> flaggedReactors = CollectionLiterals.<Object>newHashSet();
+
+    /**
+     * Return the instantiations that point to a given reactor definition.
+     * If none are known, returns an empty set. * The returned set may be
+     * unmodifiable.
+     */
+    public Set<Instantiation> getInstantiations(final Reactor definition) {
+        Set<Instantiation> instantiations = this.reactorToInstantiation.get(definition);
+        if (instantiations != null) {
+            return instantiations;
+        } else {
+            return CollectionLiterals.<Instantiation>emptySet();
+        }
+    }
+
+    /**
+     * Return the declarations that point to a given reactor definition.
+     * A declaration is either a reactor definition or an import statement.
+     */
+    public Set<ReactorDecl> getDeclarations(final Reactor definition) {
+        return this.reactorToDecl.get(definition);
+    }
+
+    /**
+     * Return the reactor definitions referenced by instantiations in this graph
+     * ordered topologically. Each reactor in the returned list is preceded by
+     * any reactors that it may instantiate.
+     */
+    public List<Reactor> getReactors() {
+        return this.nodesInTopologicalOrder();
+    }
+
+    /**
+     * Construct an instantiation graph based on the given AST and, if the
+     * detectCycles argument is true, run Tarjan's algorithm to detect cyclic
+     * dependencies between instantiations.
+     * @param resource The resource associated with the AST.
+     * @param detectCycles Whether or not to detect cycles.
+     */
+    public InstantiationGraph(final Resource resource, final boolean detectCycles) {
+        final Iterable<Instantiation> instantiations = Iterables.<Instantiation>filter(
+            IteratorExtensions.<EObject>toIterable(resource.getAllContents()), Instantiation.class);
+        final Reactor main = IterableExtensions.<Reactor>findFirst(
+            Iterables.<Reactor>filter(
+                IteratorExtensions.<EObject>toIterable(resource.getAllContents()), Reactor.class),
+            (Reactor it) -> {
+                return Boolean.valueOf((it.isMain() || it.isFederated()));
+            });
```

Comment:
I suspect that the most idiomatic way to do this whole operation (from line 116 down to here) would be to use Java's Streams API. But if we do keep using these XText library functions, then this might be more concise.
```suggestion
                (Reactor it) -> it.isMain() || it.isFederated()
            );
```

Reply:
Great point thanks!